designs/odcr.md

```diff
@@ -0,0 +1,140 @@
+# On demand capacity reservation
+
+This documents proposes supporting ODCR in Karpenter
+
+## Background
+
+In AWS [ODCR](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-capacity-reservations.html) allows users to reserve compute capacity to mitigate against the risk of being 
+unabled to get on demand capacity. This is very helpful during seasonal holidays where higher traffic is expected. 
+
+### Capacity Reservations
+
+Each Capacity Reservation is defined with:
+
+- The Availability Zone in which to reserve the capacity
+- The number of instances for which to reserve capacity
+- The instance attributes, including the instance type, tenancy, and platform/OS
+- Instance match criteria
+  - Targeted -- only accept instances that matches all attributes + explicitly targeted the capacity reservation
+  - Open -- if capacity reservation accepts all instances that matches all attributes
+
+AWS also supports grouping Capacity Reservation into Capacity Reservation groups. 
+Both these entities are supported in Launch Template's CapacityReservationTarget [definitions](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-launchtemplate-capacityreservationtarget.html).
+
+## Goals
+
+- Support associating ODCR to EC2NodeClass
+- Define Karpenter's behavior when launching nodes into Capacity Reservation
+- Define Karpenter's behavior when encountering errors when attempting to launch nodes into Capacity Reservation
+- Define Karpenter's behavior when consolidating nodes in Capacit Reservation
+
+## Non-Goals
+
+_We are keeping the scope of this design very targeted so even if these could be things we eventually support, we aren't scoping them into this design_
+- Supporting prioritization when launching nodes. We will delegate this to weights within NodePools and [default behavior](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-fleet-on-demand-backup.html#ec2-fleet-on-demand-capacity-reservations) of the CreateFleet API. 
+- Supporting changes in scaling behavior when ODCR is associated to a NodeClass. _We won't bring up N nodes to match an N node capacity reservation_
+
+## Proposed Solution
+
+### Supporting associating ODCR to EC2NodeClass
+
+Add a new field `capacityReservationSpec` to `EC2NodeClass` 
+```yaml
+apiVersion: karpenter.k8s.aws/v1beta1
+kind: EC2NodeClass
+metadata:
+  name: example-node-class
+spec:
+  capacityReservationSpec:
+    capacityReservationPreference: open | none | None  # Cannot be defined if capacityReservationTarget is specified
```

Comment:
Do we think that we need to support open ODCRs off the bat? I think the main ask here is for targeted ODCRs so I'm skeptical of using open ODCRs, particular because it becomes more difficult to model the capacity. We can definitely do it, but now you can either launch something in an ODCR or not (since you need to handle the fallback case)